libsolidity/analysis/TypeChecker.cpp

```diff
@@ -364,16 +364,24 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		}
 		else
 		{
-			if (!type(var)->canLiveOutsideStorage() && _function.isPublic())
-				m_errorReporter.typeError(3312_error, var.location(), "Type is required to live outside storage.");
+			if (
+				type(var)->containsNestedMapping() &&
+				_function.isPublic() &&
+				!isLibraryFunction
+			)
+				m_errorReporter.typeError(
+					3312_error,
+					var.location(),
+					"Types containing (nested) mappings can only be used in storage."
+				);
 			if (_function.isPublic())
 			{
 				auto iType = type(var)->interfaceType(isLibraryFunction);
 
 				if (!iType)
 				{
 					solAssert(!iType.message().empty(), "Expected detailed error message!");
-					m_errorReporter.fatalTypeError(4103_error, var.location(), iType.message());
+					m_errorReporter.typeError(4103_error, var.location(), iType.message());
```

Comment:
The error messages related to libraries happen here. This was converted to a regular type error because it was masking error messages about structs containing mappings.